### PR TITLE
Adds sounds and some small QoL improvements to the circuit imprinter, and allows them to recycle empty assemblies

### PR DESCRIPTION
**Pull request by ghost**
:cl: Xhuis
tweak: Removing and printing integrated circuits will now attempt to place them into a free hand.
tweak: You can now hit an integrated circuit printer with an unsecured electronic assembly to recycle all of the parts in the assembly en masse.
tweak: You can now recycle empty electronic assemblies in an integrated circuit printer!
soundadd: Integrated circuit printers now have sounds for printing circuits and assemblies.
/:cl:

Some low-hanging fruit for circuits. I was trying them earlier, and these were the main things that I was surprised didn't exist, so I added them. I'm not sure about the assembly manipulation yet (may have been intentional?) but it feels a lot smoother to use, especially since misprints can be emptied out and then recycled instead of awkwardly removing each component and then disposing of the case.

### CODE REVIEW

**Comment by arsserpentarium** [comment]
Thanks.
If you will make buttons to stick in gun without removing corcuit from case,it would be fabolous.

**Comment by Xhuis** [comment]
What do you mean?

File: code/modules/integrated_electronics/core/printer.dm [hunk 1]
```diff
@@ -41,9 +41,40 @@
...
+				if(!mats.has_space(mats.get_item_material_amount(IC)))
+					to_chat(user, "<span class='notice'>[src] can't hold any more materials!</span>")
+					break
+				if(!do_after(user, 5, target = user))
+					return
+				playsound(user, 'sound/items/crowbar.ogg', 50, TRUE)
+				if(EA.try_remove_component(IC, user, TRUE))
+					mats.user_insert(IC, user)
+			to_chat(user, "<span class='notice'>You recyle all the components[EA.assembly_components.len ? " you could " : " "]from [EA]!</span>")
+			playsound(user, 'sound/items/electronic_assembly_empty.ogg', 50, TRUE)
```
Comment (Cyberboss) [inline code/modules/integrated_electronics/core/printer.dm]: user -> src

File: code/modules/integrated_electronics/core/printer.dm [hunk 2]
```diff
@@ -41,9 +41,40 @@
...
+			var/only_nodrop_components = TRUE
+			for(var/V in EA.assembly_components)
+				var/obj/item/integrated_circuit/IC = V
+				if(IC.removable)
+					only_nodrop_components = FALSE
+			if(only_nodrop_components)
+				to_chat(user, "<span class='warning'>[EA] has irremovable components attached to the assembly, making it impossible to recycle!</span>")
+				return
+			to_chat(user, "<span class='notice'>You begin recycling [EA]'s components...</span>")
+			playsound(user, 'sound/items/electronic_assembly_emptying.ogg', 50, TRUE)
```
Comment (Cyberboss) [inline code/modules/integrated_electronics/core/printer.dm]: user -> src

File: code/modules/integrated_electronics/core/printer.dm [hunk 3]
```diff
@@ -172,8 +207,9 @@
 				else
 					var/datum/component/material_container/materials = GetComponent(/datum/component/material_container)
 					if(materials.use_amount_type(program["metal_cost"], MAT_METAL))
-						SScircuit.load_electronic_assembly(get_turf(src), program)
-						to_chat(usr,  "<span class='notice'>Assembly has been printed.</span>")
+						var/obj/item/assembly = SScircuit.load_electronic_assembly(get_turf(src), program)
+						to_chat(usr,  "<span class='notice'>[assembly] has been printed from the provided template!</span>")
+						playsound(usr, 'sound/items/poster_being_created.ogg', 50, TRUE)
```
Comment (Cyberboss) [inline code/modules/integrated_electronics/core/printer.dm]: usr -> src

File: code/modules/integrated_electronics/core/printer.dm [hunk 4]
```diff
@@ -126,12 +157,16 @@
...
 		var/obj/item/built = new build_type(drop_location())
+		usr.put_in_hands(built)
 
 		if(istype(built, /obj/item/device/electronic_assembly))
 			var/obj/item/device/electronic_assembly/E = built
 			E.opened = TRUE
 			E.update_icon()
 
+		to_chat(usr, "<span class='notice'>[capitalize(built.name)] printed.</span>")
+		playsound(usr, 'sound/items/jaws_pry.ogg', 50, TRUE)
```
Comment (Cyberboss) [inline code/modules/integrated_electronics/core/printer.dm]: usr -> src

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/integrated_electronics/core/printer.dm [hunk 5]
```diff
@@ -41,9 +42,50 @@
...
+				to_chat(user, "<span class='warning'>You can't reach [EA]'s components to remove them!</span>")
+				return
+			if(EA.battery)
+				to_chat(user, "<span class='warning'>Remove [EA]'s power cell first!</span>")
+				return
+			var/only_nodrop_components = TRUE
+			for(var/V in EA.assembly_components)
+				var/obj/item/integrated_circuit/IC = V
+				if(IC.removable)
+					only_nodrop_components = FALSE
```
Comment (duncathan) [inline code/modules/integrated_electronics/core/printer.dm]: break here

**Review by duncathan — DISMISSED** [review]

**Comment by arsserpentarium** [comment]
I mean to put on some modules buttons to direct interaction with installed components.Like insertion/extraction of guns.Or addition/removal liquids from storages.
Possibly I'll do it myself.But maybe you willind to help to do it.

**Comment by Xhuis** [comment]
I tried a little bit at module interactions, but I couldn't really end up coding it well. If I figure it out, I'll be happy to implement it.

File: code/modules/integrated_electronics/core/printer.dm [hunk 6]
```diff
@@ -41,9 +42,51 @@
...
+					break
+			if(only_nodrop_components)
```
Comment (Cyberboss) [inline code/modules/integrated_electronics/core/printer.dm]: You can just do this in the for loop

**Review by Cyberboss — CHANGES_REQUESTED** [review]

File: code/modules/integrated_electronics/core/printer.dm (hunk 6 above)
Comment (Xhuis) [reply code/modules/integrated_electronics/core/printer.dm]: How so?

**Review by ShizCalev — DISMISSED** [review]
sounds are fine

**Comment by Xhuis** [comment]
This has been open for a while, and it should be ready unless Cyberboss gets back to his review.

**Comment by Xhuis** [comment]
This has been up for over two weeks now. I hate to be insistent, but is there anything else I need to do in this PR for it to be mergeable?

**Pull request merged**